### PR TITLE
Use mimalloc instead of lfallocator

### PR DESCRIPTION
**Pull request by danlark1**
I hereby agree to the terms of the CLA available at: https://yandex.ru/legal/cla/?lang=en

For changelog. Remove if this is non-significant change.

Category (leave one):
- Bug Fix

Short description (up to few sentences):
Let's use mimalloc instead of lfallocator not to provide arenas for each allocation

### CODE REVIEW

File: dbms/src/Common/MiAllocator.h [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+
+#include <mimalloc.h>
+#include <cstddef>
+
+namespace DB
+{
+struct MiAllocator
+{
+
+    static void * alloc(size_t size, size_t alignment = 0)
```
Comment (alexey-milovidov) [inline dbms/src/Common/MiAllocator.h]: If you move to .cpp, you don't have to make any changes to 
`dbms/src/Interpreters/Compiler.cpp`

File: dbms/src/Common/MiAllocator.h [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+#if !USE_MIMALLOC
+#error "do not include this file until USE_MIMALLOC is set to 1"
+#endif
+
+#include <mimalloc.h>
+#include <cstddef>
+
+namespace DB
+{
+struct MiAllocator
```
Comment (alexey-milovidov) [inline dbms/src/Common/MiAllocator.h]: A comment required.

File: dbms/src/Common/MiAllocator.h (hunk 1 above)
Comment (danlark1) [reply dbms/src/Common/MiAllocator.h]: But we lose the dead code elimination property if `alignment` is zero which is rather common case

Comment (alexey-milovidov) [reply dbms/src/Common/MiAllocator.h]: Reasonable.

**Comment by alexey-milovidov** [comment]
This was not a bugfix but bug-introduce.

**Pull request merged**